modDbImpl/src/org/aion/db/impl/mockdb/PersistentMockDB.java

```diff
@@ -0,0 +1,171 @@
+ *     be useful, but WITHOUT ANY WARRANTY; without even the implied
+ *     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *     See the GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with the aion network project source files.
+ *     If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Contributors:
+ *     Aion foundation.
+ */
+package org.aion.db.impl.mockdb;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import org.aion.base.util.ByteArrayWrapper;
+
+/**
+ * Provides the same behavior as {@link MockDB} with the addition that data is read from a file on
+ * disk at open (if the file exists) and it is stored to disk at close.
+ *
+ * @author Alexandra Roatis
+ */
+public class PersistentMockDB extends MockDB {
+
+    public PersistentMockDB(String name, String path) {
+        super(name);
+
+        Objects.requireNonNull(path, "The database path cannot be null.");
+        this.path = new File(path, name).getAbsolutePath();
+    }
+
+    @Override
+    public boolean open() {
+        if (isOpen()) {
+            return true;
+        }
+
+        LOG.debug("init database {}", this.toString());
+
+        // using a regular map since synchronization is handled through the read-write lock
+        kv = new HashMap<>();
+
+        // load file from disk if it exists
+        File dbFile = new File(path);
+
+        if (dbFile.exists()) {
+            BufferedReader reader = null;
+
+            try {
+                reader = new BufferedReader(new FileReader(dbFile));
+                String text;
+
+                while ((text = reader.readLine()) != null) {
+                    String[] line = text.split(":", 2);
+                    ByteArrayWrapper key = ByteArrayWrapper.wrap(convertToByteArray(line[0]));
+                    byte[] value = convertToByteArray(line[1]);
+                    kv.put(key, value);
+                }
+            } catch (FileNotFoundException e) {
+                e.printStackTrace();
+            } catch (IOException e) {
+                e.printStackTrace();
+            } finally {
+                try {
+                    if (reader != null) {
+                        reader.close();
+                    }
+                } catch (IOException e) {
+                }
+            }
+        } else {
+            try {
+                dbFile.createNewFile();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+
+        return isOpen();
+    }
+
+    private static final byte[] convertToByteArray(String commaSeparatedNumbers) {
+        String[] numbers =
```

Comment:
The end index is exclusive in substring, not sure if this is intentional that you want to chop off the last character or not, just checking.

Reply:
I'm eliminating the `[` and `]` that are added when calling `Arrays.toString(...)` which is used for storing the data in the file.

Reply:
got it